applications: asset_tracker: Send device info and RSRP data to the cloud

hmhalvorsen opened the pull request:
The purpose of this PR is to make it possible for the asset tracker to submit device and modem information to the cloud. The commits are ordered as follows (from newest to oldest):

1. Send device info and RSRP data to the cloud.
2.  Add device info library.
3. Cloud support for device info and RSRP data.
4. Add modem_info library.
5. Add at_cmd_parser library.

This PR was originally meant to focus on the at_cmd_parser library, but has now been expanded. In order to mitigate confusion have I decided to add all of my changes which is relevant to the asset tracker here.

The end result is that the asset tracker is now able to obtain relevant device information data and send said data as one complete JSON string to the cloud. Said data is currently the modem firmware version, current link operator, and battery voltage to mention a few. The asset tracker is also available to maintain a subscription to RSRP values, which will only be updated upon change of signal strength.

Signed-off-by: Henrik Malvik Halvorsen <henrik.halvorsen@nordicsemi.no>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/NordicPlayground/fw-nrfconnect-nrf?pullRequest=471) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/NordicPlayground/fw-nrfconnect-nrf?pullRequest=471) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/NordicPlayground/fw-nrfconnect-nrf?pullRequest=471) it.</sub>

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## commit message syntax issues

3: UC4 Line exceeds max length (91>72): "The at_cmd_parser library allows parsing of received data after an AT-command transmission."

## identity/email issues

717f66ca740f00d1c0ba8516e768963047e2cbc1: author email (hmhalvorsen <henrik.halvorsen@nordicsemi.no>) needs to match one of the signed-off-by entries.
717f66ca740f00d1c0ba8516e768963047e2cbc1: author email (hmhalvorsen <henrik.halvorsen@nordicsemi.no>) does not follow the syntax: First Last <email>.

lemrey commented:
Hi, thank you for opening this PR. A few initial comments:
- who is the primary user of this library? so that I can add them as reviewers
- can you provide a sample that shows how to use this library?

hmhalvorsen commented:
I am currently working on a sample that will showcase the usage of this library.

rkertesz commented:
@hmhalvorsen Is this designed to allow someone to send an at command programmatically, parse the response, and then use that as a variable or am I not on target?

hmhalvorsen commented:
This is the intended purpose. I will hopefully have an example ready that does exactly what is asked for in the forum during the next couple of days.

lemrey commented:
Is this really a common use case that necessitates a library?

hmhalvorsen commented:
Yes, I believe so. 

The main use case is to obtain data from the modem, such as device information and the status of the LTE link. The asset tracker sample is currently being developed to obtain said data from the modem and publishing it to the cloud, and the parser library works as the engine for this case.

rkertesz commented:
@lemrey as a new user of this hardware and software I wonder if to software is the most straightforward way to perform the following..
Allow a user to be able to access network level info and use the data programmatically.  I can access temperature and accelerometer onboard in a straightforward manner but what if I want to know signal strength or other carrier info. It would be easiest for the end user to be able to encapsulate a request in something that will automatically issue the command, parse it, and return the response. This may help with that.
Honestly, the easiest solution would be some sort of way to automatically bind the command transparently. That is, in the middle of the main.c somewhere a person can say foo = getatresponse("at+cesq",6)
to get the 6th element of the response.

joakimtoe commented:
> Is this really a common use case that necessitates a library?

Yes @lemrey, I would say that most LTE applications would want at some stage to be able to get data from the modem in a easy way. @hmhalvorsen will come with a new PR showing this use soon.

lemrey commented on lib/at_cmd_parser/src/at_utils.c:
```diff
@@ -0,0 +1,45 @@
+
+#include "at_utils.h"
+
+u32_t at_params_remove_spaces_from_beginning(char **str)
```
This function does not remove spaces, it counts them.

lemrey commented on lib/at_cmd_parser/src/at_utils.c:
```diff
@@ -0,0 +1,45 @@
+	return space_count;
+}
+
+size_t at_params_get_cmd_length(const char *const str)
```
You declare the `str` parameter as a constant pointer, but then cast it and copy it to a non-constant pointer to change it. I suggest to accept a non-constant pointer to begin with, and manipulate it directly.

lemrey commented on lib/at_cmd_parser/src/at_params.c:
```diff
@@ -0,0 +1,329 @@
+#include <at_params.h>
+#include <kernel.h>
+
+#define AT_PARAMS_CALLOC k_calloc
```
Please use the OS native abstractions, without wrappers.
Let non-Zephyr code wrap Zephyr's native abstractions instead.

lemrey commented on lib/at_cmd_parser/src/at_params.c:
```diff
@@ -0,0 +1,329 @@
+		return -EINVAL;
+	}
+
+
```
Stray line

lemrey commented on lib/at_cmd_parser/src/at_cmd_parser.c:
```diff
@@ -0,0 +1,213 @@
+}
+
+
+int at_parser_max_params_from_str(const char *const at_params_str,
```
You declare `at_params_str` as constant pointer, then copy it and cast it to a non-constant one to manipulate it. I suggest to accept a non-constant pointer parameter to begin with, and get rid of the copy-casting.

metc replied:
This is by design. You are not allowed to modify the input String. This is why we use a local variable.

lemrey commented on lib/at_cmd_parser/src/at_cmd_parser.c:
```diff
@@ -0,0 +1,213 @@
+{
+	u16_t str_len;
+	bool in_double_quotes = false;
+	char *str = (char *)at_params_str;
```
See my other comments.

lemrey commented on lib/at_cmd_parser/include/at_params.h:
```diff
@@ -0,0 +1,223 @@
+ * @return Zero on success or (negative) error code otherwise.
+ */
+int at_params_put_string(const struct at_param_list *const list,
+			      size_t index, const char *const str,
```
Please check the indentation of all these declarations.

lemrey commented on lib/at_cmd_parser/src/at_cmd_parser.c:
```diff
@@ -0,0 +1,213 @@
+
+	(void)at_params_remove_spaces_from_beginning(&str);
+
+	for (size_t param_idx = 0; param_idx < max_params_count; ++param_idx) {
```
`param_idx` is an index in a for loop cycle, please rename `param_idx` to `i`, here and other places.

lemrey commented on lib/at_cmd_parser/src/at_params.c:
```diff
@@ -0,0 +1,329 @@
+#define AT_PARAMS_MALLOC k_malloc
+#define AT_PARAMS_FREE k_free
+
+/* Internal function. Parameter cannot be null. */
```
If the parameter cannot be `NULL`, then the function should either return an error or assert.
Do not rely on documentation to handle error conditions (here and other places).

metc replied:
That is wrong. This is a static (private) function. The developer of the module is responsible to check the arguments before calling this internal function, and this is the case. We do not want to null check parameters of static function because this is not efficient and not needed if the module implementation is correct.

lemrey commented on lib/at_cmd_parser/include/at_utils.h:
```diff
@@ -0,0 +1,45 @@
+ * character.
+ * @return number of removed spaces.
+ */
+u32_t at_params_remove_spaces_from_beginning(char **str);
```
Turns out this doesn't trim a string, after all; see my other comment :)

hmhalvorsen commented:
> @hmhalvorsen thank you for this, it looks good.
> I have done a first round of review and left some comments.

Thank you for your first round of review, I've taken your comments into account when updating this PR.

rkertesz commented:
Can you share some information about how to load this and use it? What needs to be done differently than the default application, on the development computer, on the nrf9160dk, and on nrfconnect online?

hmhalvorsen commented:
> Can you share some information about how to load this and use it? What needs to be done differently than the default application, on the development computer, on the nrf9160dk, and on nrfconnect online?

This PR introduces the necessary changes that are needed for the asset tracker. Simply running the asset tracker sample on a paired nRF Cloud device (such as the nRF9160 DK) will suffice. The device will submit modem info data to the cloud upon start up, and afterwards it will continuously transmit any changes to the RSRP data, if any.

hmhalvorsen commented on include/modem_info.h:
```diff
@@ -0,0 +1,75 @@
+ *
+ * @return Zero on success or (negative) error code otherwise.
+ */
+int modem_info_rsrp_sub_init(rsrp_cb_t cb);
```
Yes. 👍

hmhalvorsen commented on applications/asset_tracker/src/main.c:
```diff
@@ -372,6 +379,60 @@ static void flip_send(struct k_work *work)
+
+	sensor_data_send(&device_cloud_data);
+
+	modem_info_rsrp_sub_init(modem_rsrp_handler);
```
I agree with this statement, however, the current implementation is a workaround. 

A weakness with the modem informer is that both RSRP subscription and normal modem data acquisition share the same socket. I believe this would result in a race condition, so in order to mitigate this is subscription first initiated after the initial data transmission. (However, in the future if the cloud requests data from the modem informer would the condition be a problem again).

I suggest one of the following:

1. Keep the current implementation until a common at_cmd - module is implemented, which will solve this issue.
2. Go for your suggestion, initialize RSRP subscription upon modem_info init. (But this could lead to strange behavior without a proper fix.)
3. Develop a solid solution for said issue, however this will also be solved over time based on the work on point 1.

rkertesz commented:
Hmm. This must have some dependencies that have been updated since I've set up my environment because I am getting build errors. I don't recall them at this time however I will check. Is there documentation of the minimum version number for dependencies?

hmhalvorsen commented:
> Hmm. This must have some dependencies that have been updated since I've set up my environment because I am getting build errors. I don't recall them at this time however I will check. Is there documentation of the minimum version number for dependencies?

Are your repositories up to date? I.E zephyr/mcuboot/nrfxlib? What build errors are you getting?

hmhalvorsen commented:
> Very good! Some small fixes needed as I see it.
> 
> 1. device_info files should not be inside the asset tracker folder. Other LTE samples might also want to use this. Should be given a more descriptive name? Purpose of device_info is to format data to json?
> 2. Error in the data sent to cloud: {"BAND":20,"MODE":2,"OPERATOR":"24201","CELLID":"010EF102","IP ADDRESS":"10.160.189.83","SIM":1,"BATTERY":4188,"FW":"_nrf9160_0.���
>    Something strange happening with the FW parameter.

Thank you for taking a look at my code!

1. I've moved device_info to the lib folder now, so it will be possible for other LTE samples to use this functionality. 
The purpose of device_info is to supply an application with any data that is relevant to the device, such as link status data, current board, firmware version and so on. Right now is the only function to obtain a set of modem data and supply it to the application as a single JSON string, but this could be easily expanded in the future. I am having difficulties with finding a more descriptive name, so I am keeping the current library name as of now. I am more than open for a name change, I just haven't found any better candidates! device_info_string possibly?

2.  I have not been able to replicate your exact error, but it pointed me towards some bugs in the parser library. I have now fixed said bugs and will update this PR. Is your snippet taken from the cloud or from local terminal logging? I notice that the JSON string is too large to be fully printed in the nRF Cloud terminal. By shortening the amount of data that is written (such as removing all sensory data bar form the firmware version) I can see that the string is printed correctly. Could you try with the changes I've introduced?

rkertesz commented:
> Are your repositories up to date? I.E zephyr/mcuboot/nrfxlib? What build errors are you getting?

I used the releases as noted in the nrfconnect getting started assistant. I will update them to latest releases on master branch if the branch has been updated sine these releases.

```
cd <sourcecode_root>\ncs\zephyr ; git checkout tags/v1.13.99-ncs2
cd <sourcecode_root>\ncs\mcuboot ; git checkout tags/v1.2.99-ncs2
cd <sourcecode_root>\ncs\nrf ; git checkout tags/v0.3.0
cd <sourcecode_root>\ncs\nrfxlib ; git checkout tags/v0.3.0
```

rkertesz commented:
> I used the releases as noted in the nrfconnect getting started assistant. I will update them to latest releases on master branch if the branch has been updated sine these releases.

I updated all the dependencies and tried to build. I receive this error:
Building ‘zephyr/include/generated/driver-validation.h’ from solution ‘build’ in configuration ‘Common’
Building ‘zephyr/include/generated/kobj-types-enum.h’ from solution ‘build’ in configuration ‘Common’
Building ‘zephyr/include/generated/otype-to-str.h’ from solution ‘build’ in configuration ‘Common’
Building ‘zephyr/misc/generated/syscalls.json’ from solution ‘build’ in configuration ‘Common’
Building ‘zephyr/include/generated/syscall_list.h’ from solution ‘build’ in configuration ‘Common’
Building ‘zephyr/include/generated/syscall_dispatch.c’ from solution ‘build’ in configuration ‘Common’
Building ‘zephyr/include/generated/syscall_macros.h’ from solution ‘build’ in configuration ‘Common’
Building ‘zephyr/include/generated/offsets.h’ from solution ‘build’ in configuration ‘Common’
  Combining ‘zephyr/include/generated/offsets.h’
    
    /PROJECTCODE/ncs/zephyr/scripts/gen_offset_header.py", line 72, in <module>
    [Errno 2] No such file or directory: 'C:/PROJECTCODE/ncs/fw-nrfconnect-nrf/applications/asset_tracker/build/zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj'
Build failed

Indeed the offsets folder is empty.

lemrey commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+int modem_info_get_string(char *data_buffer, size_t valid_param, size_t max_len)
+{
+	return at_params_get_string(&m_param_list, valid_param,
+					   data_buffer, max_len);
```
Check indentation.

lemrey commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+			"%d", param_value - offset_value);
+}
+
+int modem_info_get_string(char *data_buffer, size_t valid_param, size_t max_len)
```
Simply `buf`, `len`

lemrey commented on include/device_info.h:
```diff
@@ -0,0 +1,25 @@
+ *
+ * @return Length of string buffer or (negative) error code otherwise.
+ */
+int device_info_get_string(char *string_buffer);
```
I suggest `device_info_get_string(string_buffer)` -> `device_info_string_get(buf)`

lemrey commented on lib/device_info/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+
+if(CONFIG_DEVICE_INFO)
+    zephyr_library()
+    zephyr_library_sources(device_info.c)
```
Please use the more conventional `zephyr_library_sources_ifdef` and get rid of the `if` condition.

lemrey commented on lib/device_info/device_info.c:
```diff
@@ -0,0 +1,165 @@
+		len = modem_info_update(modem_information[i]->status,
+					data_buffer);
+		if (len < 0) {
+			printk("LTE link data not obtained: %d\n", len);
```
Do not use `printk` in a library code. Use the logging subsystem instead.

lemrey commented on applications/asset_tracker/Kconfig:
```diff
@@ -286,6 +286,18 @@ config TEMP_DEV_NAME
+if MODEM_INFO
+
+config HOLD_TIME_RSRP
+	int "Minimum time between RSRP subscription messages"
```
Please specify what unit is time specified in e.g. minutes, milliseconds.

lemrey commented on applications/asset_tracker/src/main.c:
```diff
@@ -372,6 +379,60 @@ static void flip_send(struct k_work *work)
+	}
+
+	if (k_uptime_get_32() - timestamp_prev <
+	K_SECONDS(CONFIG_HOLD_TIME_RSRP)) {
```
Weird indentation.

lemrey commented on lib/device_info/device_info.c:
```diff
@@ -0,0 +1,165 @@
+	.status = MODEM_STATUS_FW_VERSION,
+};
+
+static struct modem_info *modem_information[] = {
```
These pointers can be constant.

lemrey commented on lib/device_info/device_info.c:
```diff
@@ -0,0 +1,165 @@
+	enum modem_status status;
+};
+
+static struct modem_info modem_info_rspr = {
```
Can these structs be constant?

lemrey commented on lib/at_cmd_parser/include/at_cmd_parser.h:
```diff
@@ -0,0 +1,54 @@
+/**@file at_cmd_parser.h
+ *
+ * @brief Basic parser for AT commands.
```
This is a public header, move it under `/include`.

lemrey commented on lib/at_cmd_parser/include/at_params.h:
```diff
@@ -0,0 +1,221 @@
+/**@file at_params.h
+ *
+ * @brief Store a list of AT commands/responses parameters.
```
Is this a public header? If so move it under `/include`.

lemrey commented on include/device_info.h:
```diff
@@ -0,0 +1,25 @@
+ *
+ * @brief   Device info library.
+ *
+ * Library that obtains device info data.
```
`device_info` is not a proper library name in this case, because it is too generic. This is really an LTE-specific library. We can't have a `device_info.h` public header that only applies to LTE devices.
We'll have to find a new name for this.

hmhalvorsen replied:
Yes, I agree completely.

I've done some thinking, and my solution would be the following:

1. Remove device_info completely and rather add it as a part of the modem_info library, called modem_info_json, modem_info_string or similar. 
2. There is currently a dependency to cloud in the device info library, I believe that is a poor strategy and this dependency will be removed.

jtguggedal commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+			LOG_ERR("Poll error: %d\n", err);
+			continue;
+		} else if (err == 0) {
+			LOG_DBG("Timeout");
```
Currently, we can. Timeout argument is not treated as expected in `poll()`, so we'll need to check all return values for now. Could add a TODO here to remove when `poll()` is changed.

jtguggedal commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+		err = poll(fds, nfds, K_FOREVER);
+		if (err < 0) {
+			LOG_ERR("Poll error: %d\n", err);
+			continue;
```
Underlying stuff for `poll()` is not completely implemented yet, so currently we can continue from such a situation

hmhalvorsen commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+
+static struct modem_info_data band_data = {
+	.cmd = AT_CMD_CURRENT_BAND,
+	.valid_param = MI_BAND_VALID_PARAM,
```
I've changed the name from `valid_param` to `param_index`. 

The idea is that `param_index` is the index of the parameter that we want to store from the returned AT command string. `param_count` is the number of parameters in total from the returned AT command string, this value is important for validation.

hmhalvorsen commented:
> @hmhalvorsen Thank you for addressing my comments.
> The PR has become bigger since last I checked, and I have many new comments.
> 
> Some are minor, style-related but some are more relevant; please have a look.

Thank you for another round of review! Much appreciated. 

I am currently working through your comments and I've already taken most of them into account. I will update the PR shortly when I've gone through them all.

hmhalvorsen commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+		modem_data_name[status],
+		strlen(modem_data_name[status]));
+
+	len = strlen(modem_data_name[status]);
```
It is possible by avoiding `strlen` by either implementing a clever x-macro or by simply defining the string length. I am hesitant to go for either strategy since a macro would damage the readability (and this is a public api) whilst having length defines is not that pretty in the long run in my opinion. 

I think using `strlen` here is non-optimal but is ok for this use case, as the strings themselves are short and the modem informer is not expected to execute this part of the code often (actually only on startup). 

Either way, I've moved line 240 above the `memcpy` of course, so we don't call `strlen` twice at least. :)

hmhalvorsen commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+	/* Init thread for RSRP subscription */
+	k_mutex_init(&socket_mutex);
+
+	return err ? -EFAULT : 0;
```
I am actually not sure what my reasoning was, I've removed the masking of the return value.

hmhalvorsen commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+static K_THREAD_STACK_DEFINE(socket_thread_stack, THREAD_STACK_SIZE);
+static struct k_mutex socket_mutex;
+static int at_socket_fd = INVALID_DESCRIPTOR;
+static struct pollfd fds[1];
```
To be honest, this is just a copy-paste from the at_host, I've changed this to not be an array now.

rkertesz commented:
Just to wrap up my issue, I was trying to build in Embedded studio which is not supported on master branches. I was told that I would have to build from command line instead.

ru-fu commented:
Are the doxygen comments ready for doc review? Do you have input for conceptual documentation?

rkertesz commented:
Have other folks actually successfully been able to get RSRP data to the cloud. This is all I get on connect:
I replaced any identifying info with 0000.. but you get the idea. There is no RSRP data at all but it does enterin the modem_data_init and device_stat_send functions. Something doesn't look like it is working with RSRP. Thoughts?

Application | Data
-- | --
DEVICE | {"BAND":13,"MODE":2,"OPERATOR":"0000000","CELLID":"0000000","IP ADDRESS":"00.00.00.00","SIM":1,"BATTERY":5094,"FW":"_nrf9160_0
AIR_PRESS | 98.0
TEMP | 20.5
HUMID | 51.9

hmhalvorsen commented:
> Are the doxygen comments ready for doc review? Do you have input for conceptual documentation?

hmhalvorsen commented:
> Have other folks actually successfully been able to get RSRP data to the cloud. This is all I get on connect:
> I replaced any identifying info with 0000.. but you get the idea. There is no RSRP data at all but it does enterin the modem_data_init and device_stat_send functions. Something doesn't look like it is working with RSRP. Thoughts?
> 
> Application	Data
> DEVICE	{"BAND":13,"MODE":2,"OPERATOR":"0000000","CELLID":"0000000","IP ADDRESS":"00.00.00.00","SIM":1,"BATTERY":5094,"FW":"_nrf9160_0
> AIR_PRESS	98.0
> TEMP	20.5
> HUMID	51.9
> I did some basic debugging and modem_rsrp_handler is never called, and neither is modem_info_rsrp_subscribe_thread.
> 
> It looks like modem_info_rsrp_register is called and completes without error however. I see in k_thread_create that it refers to modem_info_rsrp_subscribe_thread. Is this called correctly?
> 
> ```
> 	k_thread_create(&socket_thread, socket_thread_stack,
> 			K_THREAD_STACK_SIZEOF(socket_thread_stack),
> 			modem_info_rsrp_subscribe_thread,
> 			NULL, NULL, NULL,
> 			THREAD_PRIORITY, 0, K_NO_WAIT);
> ```

The RSRP data is not valid until the link has been up for a short while. The modem will send a signal strength notification to the application when a new unique value is present. Try to wait a while and move the device some in order to get the first initial RSRP values. Keep in mind that the application will only submit unique values (i.e. changes in the RSRP value). It has been seen that some operators actually never obtains a valid RSRP value in the first place, so the notification is never sent. Try to verify by using the at_host and sending the AT-command AT+CESQ.

It is incorrect to say that modem_rsrp_handler and modem_info_subscribe_thread is never "called", as you refer to k_thread_create which indeed creates its own thread that polls these exact functions.

hmhalvorsen commented:
> Are the doxygen comments ready for doc review? Do you have input for conceptual documentation?

I believe the doxygen comments are ready for review. I could start working on some documentation and send you the draft?

evenl commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+	return len <= 0 ? -EINVAL : len;
+}
+
+int modem_info_short_get(char *out, size_t param_index,
```
This is an internal function? Should be a static function. Do you need to wrap this function?

evenl commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+		return err;
+	}
+
+	return snprintf(out, len,
```
Is this the place to turn the value back into a string? Maybe it should be stored as a number until you actually know it will be used in JSON formatted data. 

Seems like the `modem_info_json_string_get` is anyway turning the string back to into a number while creating the JSON formatted data object.

hmhalvorsen replied:
I've done some thinking. 

The asset tracker requires the data to be outputted to the cloud to be sent as a string. There is currently no information about the parameter type on the application level, and for the case of the asset tracker is it not desired to add any parameter type check logic either. This means there must be some way to obtain a int/short value from the modem informer as a string. 

However, it is also desirable as you describe to use the short/int values as-is, such as for battery voltage checks or some sort of temperature logic or whatever. 

I believe the best solution would be to make `modem_info_update` return the parameter value as either a string or an integer by request from the application. (Of course, requesting a string as a number value should not be allowed). My way of doing that would be to add another input argument to `modem_info_update`, which would be the requested data type, and perform logic on this argument. The input buffer would then have to change from `char` to `void`, which I am conflicted about. What are your thoughts about this solution @evenl ?

evenl replied:
I don't think it's the modem_info_update functions responsibility to convert the data into any other type because the modem_info_update function should not know anything about what a potential caller want. The conversion should happen in the `modem_info_json.c,` where the information of which type is needed resides. 

There is a software principle called "Single Responsibility Principle" which I think apply in this situation.

hmhalvorsen replied:
All right, I will try a different strategy. I think the modem_info should simply have two public functions then; one for returning a parameter as a string and one for returning the parameter as a short/int value, if applicable. I will update the PR when this is done.

evenl commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+
+	k_mutex_unlock(&socket_mutex);
+
+	return 0;
```
Every AT command will return 'OK', 'ERROR', '+CMS', or '+CME' you should verify that you get an 'OK' back before you return 0 back to the caller.

hmhalvorsen replied:
All right, I will implement this.

evenl commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+		modem_data_name[info],
+		len);
+
+	return len <= 0 ? -EINVAL : len;
```
Shouldn't you do this check before you do a memcpy of the data?

hmhalvorsen replied:
I agree.

evenl commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+	u32_t param_index;
+
+	err = at_parser_max_params_from_str(
+		&buf[CMD_SIZE(modem_data->cmd)], &m_param_list,
```
The command/data type should be item 0 in the parameter list returned from the at_parser

(I know it's not your code, just checking if people agree with me)

hmhalvorsen replied:
If this is required I guess this could be implemented in the future.

evenl commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+		return err;
+	}
+
+	err = modem_info_parse(modem_data[info], recv_buf);
```
How do you know that you actually received the response for the command you just issued and not for instance a CESQ notification or some other notification?

hmhalvorsen replied:
That is currently the largest weakness of the modem informer. The only solution I've come up with as of now is what you are currently working on with your library. Is there another (temporary) solution, perhaps?

hmhalvorsen replied:
After some thinking I've decided to check if the received message equals to an expected response of that command, I believe this is feasible.

hmhalvorsen replied:
Since your project is addressing this problem have I decided to keep this as is to avoid overlapping.

hmhalvorsen commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+
+int modem_info_string_get(char *buf, size_t param_index, size_t len)
+{
+	return at_params_get_string(&m_param_list, param_index,
```
My mistake, I will update this to follow my convention `object_verb`.

hmhalvorsen commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,412 @@
+				CONFIG_MODEM_INFO_MAX_AT_PARAMS_RSP);
+
+	/* Occupy socket for AT command transmission */
+	at_socket_fd = socket(AF_LTE, 0, NPROTO_AT);
```
I am aware of this library, I am sure the modem informer would benefit greatly from such a library. Please let me know when the library is nearing completion.

hmhalvorsen commented on include/modem_info.h:
```diff
@@ -0,0 +1,84 @@
+};
+
+/**@brief LTE link data type. */
+enum modem_data_type {
```
Nice catch. I will remove this enum.

rkertesz commented:
> It has been seen that some operators actually never obtains a valid RSRP value in the first place, so the notification is never sent. Try to verify by using the at_host and sending the AT-command AT+CESQ.

I tested with at commands and am getting rsrp numbers between 42 and 47 but, try as I might, I cannot get this nrf asset tracker application to report a change is rsrq, no matter where I go. Hmm. Is there a straightforward way I can poll cesq every minute or hijack the sensor polling to include rsrq (although 2 seconds is too fast)?

hmhalvorsen commented:
> > It has been seen that some operators actually never obtains a valid RSRP value in the first place, so the notification is never sent. Try to verify by using the at_host and sending the AT-command AT+CESQ.
> 
> I tested with at commands and am getting rsrp numbers between 42 and 47 but, try as I might, I cannot get this nrf asset tracker application to report a change is rsrq, no matter where I go. Hmm. Is there a straightforward way I can poll cesq every minute or hijack the sensor polling to include rsrq (although 2 seconds is too fast)?

I am currently undergoing some changes to this PR. I will verify that RSRP subscription works and come back to you. In the meantime: have you tried adjusting HOLD_TIME_RSRP in the asset tracker's kconfig? It seems to me you desire a quicker update rate.

hmhalvorsen commented:
> > It has been seen that some operators actually never obtains a valid RSRP value in the first place, so the notification is never sent. Try to verify by using the at_host and sending the AT-command AT+CESQ.
> 
> I tested with at commands and am getting rsrp numbers between 42 and 47 but, try as I might, I cannot get this nrf asset tracker application to report a change is rsrq, no matter where I go. Hmm. Is there a straightforward way I can poll cesq every minute or hijack the sensor polling to include rsrq (although 2 seconds is too fast)?

@rkertesz I were able to detect a couple of bugs related to how the subscription thread was handled. Said bugs would make it difficult for you to obtain RSRP data. Please try the updated code at your convenience and report back to me.

rkertesz commented:
> @rkertesz I were able to detect a couple of bugs related to how the subscription thread was handled. Said bugs would make it difficult for you to obtain RSRP data. Please try the updated code at your convenience and report back to me.

Thank you. It looks like it is working better now. It does look the RSRP data are getting posted now. Perhaps the signal isn't changing much between my test locations right now but it is very infrequent for RSRP data to post, meaning that if I go back and forth between locations, sometimes it updates and sometimes it does not. I updated the frequency to a very short 10 seconds by the way. I may look into how to trigger the request on a timer instead of subscribing to changes. This fits with my use case better because the payload should have the RSRP in it when I am sending data.

hmhalvorsen commented:
> > @rkertesz I were able to detect a couple of bugs related to how the subscription thread was handled. Said bugs would make it difficult for you to obtain RSRP data. Please try the updated code at your convenience and report back to me.
> 
> Thank you. It looks like it is working better now. It does look the RSRP data are getting posted now. Perhaps the signal isn't changing much between my test locations right now but it is very infrequent for RSRP data to post, meaning that if I go back and forth between locations, sometimes it updates and sometimes it does not. I updated the frequency to a very short 10 seconds by the way. I may look into how to trigger the request on a timer instead of subscribing to changes. This fits with my use case better because the payload should have the RSRP in it when I am sending data.

Very good. Try calling modem_info_short_get(MODEM_INFO_RSRP, your_data_buffer), it should return the RSRP value on demand.

lemrey commented on lib/modem_info/modem_info_json.c:
```diff
@@ -0,0 +1,157 @@
+			ret += json_add_num(data_obj, data_name, param_value);
+		}
+
+		memset(data_name, 0, MODEM_INFO_MAX_RESPONSE_SIZE);
```
Just a very minor observation: if you keep this `memset` closer to where you fill the buffer it's easier to see that the buffer is cleared each time before it is used, and you could probably get rid of one extra `memset` (i.e. avoid zeroing the buffer on declaration).

hmhalvorsen replied:
All right, I've moved it.

lemrey commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,444 @@
+	if (resp_buffer != NULL) {
+		len = recv(fd, resp_buffer, CONFIG_MODEM_INFO_BUFFER_SIZE, 0);
+
+		if ((len < CMD_SIZE(success)+1) ||
```
Simplify to `strlen()` to avoid the `+1`

hmhalvorsen replied:
Fixed

ru-fu commented on include/at_params.h:
```diff
@@ -0,0 +1,253 @@
+ *
+ * @param[in] list    Parameter list.
+ * @param[in] index   Parameter index in the list.
+ * @param[out] len    Length of the parameter in bytes or the string length.
```
What does the "or" mean here? 
Can it return either or, or are we saying the same thing in two different ways?

hmhalvorsen replied:
We are saying the same thing in two different ways here. It would be sufficient to remove "or the string length" at the end of the sentence.

ru-fu replied:
fixed

ru-fu commented on include/modem_info.h:
```diff
@@ -0,0 +1,123 @@
+ */
+int modem_info_short_get(enum modem_info info, u16_t *buf);
+
+/** @brief Function for requesting the string name of a modem
```
What is meant with "string name"?

hmhalvorsen replied:
This is a mistake. It is the name of the modem data type that is requested, not a "string name", whatever that is. It will be returned as a string however.

ru-fu replied:
fixed

ru-fu commented on lib/modem_info/Kconfig:
```diff
@@ -0,0 +1,40 @@
+#
+
+menuconfig MODEM_INFO
+	bool "nRF91 Modem info library"
```
```suggestion
	bool "nRF91 modem information library"
```

ru-fu commented on lib/at_cmd_parser/include/at_utils.h:
```diff
@@ -0,0 +1,45 @@
+/**@file at_utils.h
```
Do we need documentation for this file?

hmhalvorsen replied:
I guess not, this is not a public API.

ru-fu replied:
OK. Should it be in the include folder then? (I have no idea ;) )

ru-fu commented on lib/modem_info/Kconfig:
```diff
@@ -0,0 +1,40 @@
+	int "Size of buffer used to read data from the socket"
+	default 128
+	help
+		Set the buffer size which contains the returned
```
```suggestion
		Set the size of the buffer that contains the returned

ru-fu commented on lib/modem_info/Kconfig:
```diff
@@ -0,0 +1,40 @@
+	default 128
+	help
+		Set the buffer size which contains the returned
+		string after an AT-command. The buffer is processed
```
```suggestion
		string after an AT command. The buffer is processed

hmhalvorsen commented:
> I have added a commit with documentation - please review.
> 
> Some questions (in addition to those I added for specific files):
> 
> * We now have 3 libraries related to AT commands. It would be nice to group them in the output - but the only sensible way to do this is if they are grouped in the input as well. Would it make sense to add them in an at_commands folder under include?
> * Are there any updates needed to the asset tracker documentation?
>   https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/applications/asset_tracker/README.html

Thank you for your assistance @ru-fu ! Much appreciated.

* I believe there is a general consensus that the include folder itself is starting to get a bit messy. There should be some housecleaning there, but I don't think the way going forward would be to introduce subfolders based on categories such as AT-command functionality and similar. I think we should leave it as is for now. CC @joakimtoe 
* The asset tracker now supports RSRP subscription and different device data defined in the modem_info documentation (such as battery voltage and current operator, all under the umbrella term "DEVICE" from the cloud's perspective). An older update also introduced support for humidity data and air pressure (HUMID and AIR_PRESS), I don't see that information in the documentation. This PR also introduces new dependencies to the libraries modem_info and at_cmd_parser.

Your changes looks very good to me. Also: should I update this PR with your comments or are you able to?

ru-fu commented:
Thanks - I updated the documentation, please have another look.

Will you do the changes to the Kconfig file?

hmhalvorsen commented:
> Thanks - I updated the documentation, please have another look.
> 
> Will you do the changes to the Kconfig file?

Great, I will update the Kconfig file shortly.

hmhalvorsen commented:
@ru-fu I've updated the Kconfig now. Your newest commit looks good to me.

lemrey commented on lib/at_cmd_parser/include/at_utils.h:
```diff
@@ -0,0 +1,45 @@
+
+#include <stdint.h>
+
+#include "compiler_abstraction.h"
```
This `#include` is not needed in this header.

hmhalvorsen replied:
Fixed

lemrey commented on lib/at_cmd_parser/include/at_utils.h:
```diff
@@ -0,0 +1,45 @@
+#ifndef AT_UTILS_H_
+#define AT_UTILS_H_
+
+#include <stdint.h>
```
`u32_t` is not defined in `stdint.h`. Include `zephyr/types.h` instead.

hmhalvorsen replied:
Fixed

lemrey commented on include/at_params.h:
```diff
@@ -0,0 +1,221 @@
+#ifndef AT_PARAMS_H_
+#define AT_PARAMS_H_
+
+#include <stdint.h>
```
`u32_t`, `u16_t` etc are not defined in `stdint.h`. Include `zephyr/types.h` instead.

hmhalvorsen replied:
Fixed

lemrey commented on include/at_cmd_parser.h:
```diff
@@ -0,0 +1,54 @@
+#ifndef AT_CMD_PARSER_H_
+#define AT_CMD_PARSER_H_
+
+#include <zephyr.h>
```
Why do you include this header? Could you include `zephyr/types.h` instead?

hmhalvorsen replied:
Fixed

lemrey commented on lib/at_cmd_parser/src/at_params.c:
```diff
@@ -0,0 +1,327 @@
+		return -EACCES;
+	}
+
+	list->params = k_calloc(max_params_count,
```
Could you please check if you agree and if so remove the `for` loop below?

hmhalvorsen replied:
I agree, I've removed the for loop. Nice bit of information, thank you for sharing. :)

metc replied:
Each parameter will be initialized to `AT_PARAM_TYPE_EMPTY` using `calloc`. This was the original design.

hmhalvorsen commented on lib/at_cmd_parser/src/at_params.c:
```diff
@@ -0,0 +1,327 @@
+		return -EINVAL;
+	}
+
+	char *param_value = (char *)k_malloc(str_len + 1);
```
I didn't know, thank you.

hmhalvorsen commented:
I've updated the PR with all minor comments that were left out. @evenl , could you please recheck?

evenl commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,444 @@
+	if (resp_buffer != NULL) {
+		len = recv(fd, resp_buffer, CONFIG_MODEM_INFO_BUFFER_SIZE, 0);
+
+		if ((len < strlen(success)) ||
```
Wouldn't memcmp return != 0 in this case also?

hmhalvorsen replied:
Yes, of course. I'll remove the length check.

evenl commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,444 @@
+		}
+	}
+
+	k_mutex_unlock(&socket_mutex);
```
There are multiple occurrences where you return before you unlock the mutex. You should probably store any error code in a variable so that your execution path always passes this point, and then return the variable at the bottom.

hmhalvorsen replied:
Fixed.

evenl commented on lib/modem_info/modem_info.c:
```diff
@@ -0,0 +1,444 @@
+
+static bool is_cesq_notification(char *buf, size_t len)
+{
+	if ((len < CMD_SIZE(AT_CMD_CESQ_RESP)) ||
```
Couldn't you just use memcmp here instead? I am not sure that this expression will always work as you expect it to.

hmhalvorsen replied:
I've decided to remove the if-check entirely and go directly to the strstr call.

hmhalvorsen commented:
I've updated the PR with your comments @evenl , thank you for your input.

lemrey commented:
@hmhalvorsen I think we should squash bd7627135cd0d33905e389140426d5d4a18efe57 into the commit that makes the change in the code. @ru-fu ?

ru-fu commented:
> 
> 
> @hmhalvorsen I think we should squash [bd76271](https://github.com/NordicPlayground/fw-nrfconnect-nrf/commit/bd7627135cd0d33905e389140426d5d4a18efe57) into the commit that makes the change in the code. @ru-fu ?

We can, but it can also be a commit on its own (since it's not a fix to the doc update, but the actual update). I don't care. ;)

metc commented on lib/at_cmd_parser/include/at_cmd_parser.h:
```diff
@@ -0,0 +1,54 @@
+ *
+ * @return Zero on success or (negative) error code otherwise.
+ */
+int at_parser_max_params_from_str(const char *const at_params_str,
```
In my opinion, `const` keywords are important here. They mean the pointer or String value will not be touch by this function. This is not a pass-by-value parameter.

metc commented on lib/at_cmd_parser/src/at_cmd_parser.c:
```diff
@@ -0,0 +1,213 @@
+	u32_t value = 0;
+	u32_t check_value;
+	bool negative = false;
+	const char *at_str = at_params_str;
```
The point is that you should never change the input parameter, this is why we used a local pointer. There is no copy here.
